morl_baselines/multi_policy/pcn/pcn.py

```diff
@@ -0,0 +1,441 @@
+        self.hidden_dim = hidden_dim
+        self.scaling_factor = scaling_factor
+        self.desired_return = None
+        self.desired_horizon = None
+
+        self.model = Model(
+            self.observation_dim, self.action_dim, self.reward_dim, self.scaling_factor, hidden_dim=self.hidden_dim
+        ).to(self.device)
+        self.opt = th.optim.Adam(self.model.parameters(), lr=self.learning_rate)
+
+        self.log = log
+        if log:
+            self.setup_wandb(project_name, experiment_name)
+
+    def get_config(self) -> dict:
+        """Get configuration of PCN model."""
+        return {
+            "batch_size": self.batch_size,
+            "gamma": self.gamma,
+            "learning_rate": self.learning_rate,
+            "hidden_dim": self.hidden_dim,
+            "scaling_factor": self.scaling_factor,
+        }
+
+    def update(self) -> None:
+        """Update PCN model."""
+        batch = []
+        # randomly choose episodes from experience buffer
+        s_i = np.random.choice(np.arange(len(self.experience_replay)), size=self.batch_size, replace=True)
+        for i in s_i:
+            # episode is tuple (return, transitions)
+            ep = self.experience_replay[i][2]
+            # choose random timestep from episode,
+            # use it's return and leftover timesteps as desired return and horizon
+            t = np.random.randint(0, len(ep))
+            # reward contains return until end of episode
+            s_t, a_t, r_t, h_t = ep[t].observation, ep[t].action, np.float32(ep[t].reward), np.float32(len(ep) - t)
+            batch.append((s_t, a_t, r_t, h_t))
+
+        obs, actions, desired_return, desired_horizon = zip(*batch)
+        log_prob = self.model(
+            th.tensor(obs).to(self.device),
+            th.tensor(desired_return).to(self.device),
+            th.tensor(desired_horizon).unsqueeze(1).to(self.device),
+        )
+
+        self.opt.zero_grad()
+        # one-hot of action for CE loss
+        actions = F.one_hot(th.tensor(actions).long().to(self.device), len(log_prob[0]))
+        # cross-entropy loss
+        l = th.sum(-actions * log_prob, -1)
+        l = l.mean()
+        l.backward()
+        self.opt.step()
+
+        return l, log_prob
+
+    def _add_episode(self, transitions, max_size: int, step: int) -> None:
+        # compute return
+        for i in reversed(range(len(transitions) - 1)):
+            transitions[i].reward += self.gamma * transitions[i + 1].reward
+        # pop smallest episode of heap if full, add new episode
+        # heap is sorted by negative distance, (updated in nlargest)
+        # put positive number to ensure that new item stays in the heap
+        if len(self.experience_replay) == max_size:
+            heapq.heappushpop(self.experience_replay, (1, step, transitions))
+        else:
+            heapq.heappush(self.experience_replay, (1, step, transitions))
+
+    def _nlargest(self, n, threshold=0.2):
```

Comment:
This is explained in the paper, maybe add in the doc for people to check at the paper for the buffer storage criterion.